mbox nxp imx mu

mjchen0 opened the pull request:
Ignore TriggerInterrupt failures. A failure is returned by the SDK hal driver if the interrupt is already pending, but returning an error up is problematic for certain clients, particularly icmsg.c. icmsg uses mailbox for notifications/interrupts, and doesn't expect any errors. In some places it propagates an error up, in some places it ignores errors, and in some places it asserts. The asserts in particular are fatal.

As long as the interrupt is pended, we should be able to ignore the case if the interrupt is already pended
and not consider that an error, since the interrupt handler on the remote processor should read all pending data.

Also make some optimizations to interrupt handling:
* remove extra trampoline layer to get the device
* return from the ISR early when all interrupts have been handled

mjchen0 committed:
drivers: mbox: imx_mu: ignore TriggerInterrupt failures

The ipc backends, like icmsg.c, don't expect errors when
using mailbox send for notifications. In some places,
icmsg.c ignores the return value of the mailbox send,
in some places it propagates the error up, and in some
places it asserts.

icmsg only uses the mailbox send with NULL msg as a means
to interrupt the other processor, and all data is passed
via shared memory. So it is safe to ignore the case when
the interrupt is already pending, since the other
processor should handle the pending message when it gets
around the handling the interrupt. As long as the order
of clearing the interrupt is before the handling of the
data, there should be no case where not pending a new
interrupt when one is already pending would result in missing
any new message.

Signed-off-by: Mike J. Chen <mjchen@google.com>

iuliana-prodan commented on drivers/mbox/mbox_nxp_imx_mu.c:
```diff
@@ -151,50 +151,57 @@ static DEVICE_API(mbox, nxp_imx_mu_driver_api) = {
 	.set_enabled = nxp_imx_mu_set_enabled,
```
I agree with @dbaluta.
Please split this in, at least, 3 patches.

mjchen0 replied:
Okay, 2nd commit split now into 4 patches.

iuliana-prodan commented on drivers/mbox/mbox_nxp_imx_mu.c:
```diff
@@ -151,50 +151,57 @@ static DEVICE_API(mbox, nxp_imx_mu_driver_api) = {
 
 	for (int i_channel = 0; i_channel < MU_MAX_CHANNELS; i_channel++) {
-		const uint32_t rx_int_mask = g_rx_flag_mask[i_channel];
+		/* handle notification interrupts first because they are
```
This could break existing assumptions.
You update the channel priority order here: previously, channel 0 data had higher priority than channel 1 notifications. Now, channel 7 notifications have higher priority than channel 0 data.

A fix for this is:
```
    for (int i_channel = 0; i_channel < MU_MAX_CHANNELS && flags != 0; i_channel++) {
        const uint32_t gen_int_mask = g_gen_int_pend_mask[i_channel];
        const uint32_t rx_int_mask = g_rx_flag_mask[i_channel];
        
        // handle notification first for current channel
        ...
        // then handle data for current channel
        ...
```

mjchen0 replied:
You brought up an interesting point and I thought about it for a while. I think it was actually a bug in the previous implementation that if the data interrupt was pending, it would skip the notification interrupt. The ISR would have to fire again to handle the notification interrupt. Even worse, if a new data pending interrupt arrived before the ISR fired, the notification interrupt would again be skipped. That's not a common usage case I believe (I only have used the imx_mu for icbmsg, which only uses the notification interrupt, so there would never be both interrupts pending for the channel), but I think it's still a bug, so I put a patch in to fix that first, and then another patch to change the order.

I believe my patches don't change the priority in that the lowest channel number is still highest priority, though now it does make it such that both interrupts of the lowest channels are checked before any interrupts of the next channel. Please review again and confirm I didn't miss anything. Thanks!.

iuliana-prodan replied:
I believe the new implementation is more efficient and robust. Thanks @mjchen0!

iuliana-prodan reviewed (changes requested):
Just a couple of small concerns, but overall LGTM.

LE: Sorry for the late review, I totally missed this PR :(

dleach02 commented:
@mjchen0 will you be able to address the comments?

mjchen0 committed:
drivers: mbox: imx_mu: Remove extra interrupt handler layer

Change to pass the device node to IRQ_CONNECT and remove
the intermediate interrupt handler that has to call
DEVICE_DT_INST_GET() to get the device pointer.

Signed-off-by: Mike J. Chen <mjchen@google.com>

mjchen0 committed:
drivers: mbox: imx_mu: handle all interrupts for a channel

Handle all interrupts for a channel before going to next
channel. Previous code did an if/else check, so if both
a data rx interrupt and a notification interrupt were pending
for a channel, only one would be handled in the ISR. The
ISR would have to fire again (probably would be still pending)
for the other interrupt to be processed. This could even
lead to starvation/delay if another data rx interrupt got
pended again.

Signed-off-by: Mike J. Chen <mjchen@google.com>

mjchen0 committed:
drivers: mbox: imx_mu: change order of interrupt handling

Handle notification interrupt before data interrupt of
a channel. Notification interrupts are more commonly
used (e.g. for ipc) and clearing the interrupt ASAP
helps reduce the possiblility that the sender finds
the interrupt already pending if it tries to send multiple
ipc messages in quick succession.

Signed-off-by: Mike J. Chen <mjchen@google.com>

mjchen0 committed:
drivers: mbox: imx_mu: return early in ISR when done

Change the ISR code to return once all pended interrupts
are handled instead of continuing to iterate over
all channels. Common occurence is just for one interrupt
bit to be set so no need to check all bits in all channels
if only one bit is pending.

Signed-off-by: Mike J. Chen <mjchen@google.com>
